File: packages/rslint-test-tools/tests/eslint/rules/no-duplicate-case.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-duplicate-case', {
  valid: [
    'switch (a) { case 1: break; case 2: break; }',
    'switch (a) { case 1: break; case "1": break; }',
    'switch (a) { case 1: break; default: break; }',
    'switch (a) { case "a": break; case "b": break; }',
    'switch (a) { case a: break; case b: break; }',
    // String literals with comment-like content should not be corrupted
    'switch (a) { case "http://example.com": break; case "other": break; }',
    'switch (a) { case "a /* b */": break; case "a": break; }',
    // String literals with different whitespace should not be collapsed
    'switch (a) { case "hello  world": break; case "hello world": break; }',
    // Regex literals should not be confused with comments
    'switch (a) { case /foo/: break; case /bar/: break; }',
    'switch (a) { case /foo/i: break; case /bar/g: break; }',
  ],
  invalid: [
    {
      code: 'switch (a) { case 1: break; case 1: break; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'switch (a) { case "a": break; case "a": break; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'switch (a) { case 1: break; case 2: break; case 1: break; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'switch (a) { case a: break; case a: break; }',
      errors: [{ messageId: 'unexpected' }],
    },
    // Regex containing /* in character class should not be confused with comment
    {
      code: 'switch (a) { case /[/*]/: break; case /[/*]/: break; }',
      errors: [{ messageId: 'unexpected' }],
    },
    // Comments outside strings should still be stripped for comparison
    {
      code: 'switch (a) { case /*a*/ 1: break; case 1: break; }',
      errors: [{ messageId: 'unexpected' }],
    },
  ],
});

File: internal/rules/no_duplicate_case/no_duplicate_case_test.go
package no_duplicate_case

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoDuplicateCaseRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&NoDuplicateCaseRule,
		[]rule_tester.ValidTestCase{
			{Code: `switch (a) { case 1: break; case 2: break; }`},
			{Code: `switch (a) { case 1: break; case "1": break; }`},
			{Code: `switch (a) { case 1: break; default: break; }`},
			{Code: `switch (a) { case "a": break; case "b": break; }`},
			{Code: `switch (a) { case a: break; case b: break; }`},
			// String literals with comment-like content should not be corrupted
			{Code: `switch (a) { case "http://example.com": break; case "other": break; }`},
			{Code: `switch (a) { case "a /* b */": break; case "a": break; }`},
			// String literals with different whitespace should not be collapsed
			{Code: `switch (a) { case "hello  world": break; case "hello world": break; }`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `switch (a) { case 1: break; case 1: break; }`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 29},
				},
			},
			{
				Code: `switch (a) { case "a": break; case "a": break; }`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 31},
				},
			},
			{
				Code: `switch (a) { case 1: break; case 2: break; case 1: break; }`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 44},
				},
			},
			{
				Code: `switch (a) { case a: break; case a: break; }`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 29},
				},
			},
			// Comments outside strings should still be stripped for comparison
			{
				Code: `switch (a) { case /*a*/ 1: break; case 1: break; }`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 35},
				},
			},
		},
	)
}
